features/storage/kvstore/filesystemstore/FileSystemStore.cpp

```diff
@@ -0,0 +1,573 @@
+    strncpy(set_handle->key, key, key_len);
+    set_handle->key[key_len] = '\0';
+    *handle = (set_handle_t)set_handle;
+
+    key_metadata.magic = FSST_MAGIC;
+    key_metadata.metadata_size = sizeof(key_metadata_t);
+    key_metadata.revision = FSST_REVISION;
+    key_metadata.user_flags = create_flags;
+    kv_file.write(&key_metadata, sizeof(key_metadata_t));
+    kv_file.close();
+exit_point:
+    if (status != KVSTORE_SUCCESS) {
+        _mutex.unlock();
+    }
+    return status;
+}
+
+int FileSystemStore::set_add_data(set_handle_t handle, const void *value_data, size_t data_size)
+{
+    int status = KVSTORE_SUCCESS;
+    size_t added_data = 0;
+    inc_set_handle_t *set_handle = (inc_set_handle_t *)handle;
+    File kv_file;
+
+    if ((value_data == NULL) || (handle == NULL) ) {
+        status = KVSTORE_BAD_VALUE;
+        goto exit_point;
+    }
+
+    if ( (_cur_inc_data_size + data_size) > set_handle->data_size ) {
+        tr_error("Added Data exceeds set_start final size - corrupted data, deleteing file: %s", _full_path_key);
+        _fs->remove(_full_path_key);
+        status = KVSTORE_DATA_CORRUPT;
+        goto exit_point;
+    }
+
+    if (0 != kv_file.open(_fs, _full_path_key, O_WRONLY | O_APPEND)) {
+        status = KVSTORE_NOT_FOUND;
+        goto exit_point;
+    }
+
+    added_data = kv_file.write(value_data, data_size);
+    if (added_data != data_size) {
+        status = KVSTORE_OS_ERROR;
+    }
+    _cur_inc_data_size += added_data;
+
+    kv_file.close();
+
+exit_point:
+
+    return status;
+
+}
+
+int FileSystemStore::set_finalize(set_handle_t handle)
+{
+    int status = KVSTORE_SUCCESS;
+    inc_set_handle_t *set_handle = NULL;
+
+    if (handle == NULL) {
+        status =  KVSTORE_BAD_VALUE;
+        goto exit_point;
+    }
+
+    set_handle = (inc_set_handle_t *)handle;
+
+    if (set_handle->key == NULL) {
+        status = KVSTORE_DATA_CORRUPT;
+    } else {
+        if (_cur_inc_data_size != set_handle->data_size ) {
+            tr_error("Added Data doesn't match set_start final size - corrupted data, deleteing file: %s", _full_path_key);
+            _fs->remove(_full_path_key);
+            status = KVSTORE_DATA_CORRUPT;
+        }
+
+        delete set_handle->key;
+    }
+
+    delete set_handle;
+    _cur_inc_data_size = 0;
+
+exit_point:
+    if (status != KVSTORE_BAD_VALUE) {
+        _mutex.unlock();
+    }
+
+    return status;
+}
+
+int FileSystemStore::iterator_open(iterator_t *it, const char *prefix)
+{
+    int status = KVSTORE_SUCCESS;
+    Dir *kv_dir = NULL;
+    key_iterator_handle_t *key_it = NULL;
+
+    if (it == NULL) {
+        return KVSTORE_BAD_VALUE;
+    }
+
+    _mutex.lock();
+    if (false == _is_initialized) {
+        status = KVSTORE_UNINITIALIZED;
+        goto exit_point;
+    }
+    key_it = new key_iterator_handle_t;
+    key_it->dir_handle = NULL;
+    key_it->prefix = NULL;
+    if (prefix != NULL) {
+        key_it->prefix = new char[KVStore::MAX_KEY_SIZE + 1];
```

Comment:
You can use `strdup` instead of these three lines. It does just the same (only with the actual size). You just need to free it with `free` and not with `delete` when you're done.

Reply:
Thanks, changed to strndup...